src/Database/LSMTree/Internal/BloomFilterQuery2.hs

```diff
@@ -0,0 +1,441 @@
+-- memory. It also provides an opportunity to share key hashes across filters.
+--
+-- We have as inputs N bloom filters and M keys to look up in them. So overall
+-- there are N * M independent bloom filter tests. The result is expected to be
+-- sparse, with roughly M*FPR positive results. So we use a sparse
+-- representation for the result: pairs of indexes identifying input Bloom
+-- filters and input keys with a positive test result.
+--
+-- We aim for the number of memory operations in-flight simultaneously to be on
+-- the order of 32. This trades-off memory parallelism with cache use. In
+-- particular this means the algorithm must be able to use a fixed prefetch
+-- \"distance\" rather than this being dependent on the input sizes. To achieve
+-- this, we use a fixed size circular buffer (queue). The buffer size can be
+-- tuned to optimise the prefetch behaviour: indeed we pick exactly 32.
+--
+-- Micro optimisation
+-- ------------------
+--
+-- We use primitive arrays and arrays -- rather than vectors -- so as to
+-- minimise the number of function variables, to be able to keep most things in
+-- registers. Normal vectors use two additional variables, which triples
+-- register pressure.
+--
+-- We use an Array of Strict BloomFilter. This avoids the test for WHNF in the
+-- inner loop, which causes all registered to be spilled to and restored from
+--the stack.
+--
+
+-- | A 'RunIxKeyIx' is a (compact) pair of a 'RunIx' and a 'KeyIx'.
+--
+-- We represent it as a 32bit word, using:
+--
+-- * 16 bits for the run\/filter index (MSB)
+-- * 16 bits for the key index (LSB)
+--
+newtype RunIxKeyIx = MkRunIxKeyIx Word32
+  deriving stock Eq
+  deriving newtype P.Prim
+
+pattern RunIxKeyIx :: RunIx -> KeyIx -> RunIxKeyIx
+pattern RunIxKeyIx r k <- (unpackRunIxKeyIx -> (r, k))
+  where
+    RunIxKeyIx r k = packRunIxKeyIx r k
+{-# INLINE RunIxKeyIx #-}
+{-# COMPLETE RunIxKeyIx #-}
+
+packRunIxKeyIx :: Int -> Int -> RunIxKeyIx
+packRunIxKeyIx r k =
+    assert (r >= 0 && r <= 0xffff
+         && k >= 0 && k <= 0xffff) $
+    MkRunIxKeyIx $
+      (fromIntegral :: Word -> Word32) $
+        (fromIntegral r `unsafeShiftL` 16)
+     .|. fromIntegral k
+{-# INLINE packRunIxKeyIx #-}
+
+unpackRunIxKeyIx :: RunIxKeyIx -> (Int, Int)
+unpackRunIxKeyIx (MkRunIxKeyIx c) =
+    ( fromIntegral (c `unsafeShiftR` 16)
+    , fromIntegral (c .&. 0xfff)
+    )
+{-# INLINE unpackRunIxKeyIx #-}
+
+instance Show RunIxKeyIx where
+  showsPrec _ (RunIxKeyIx r k) =
+    showString "RunIxKeyIx " . showsPrec 11 r
+              . showChar ' ' . showsPrec 11 k
+
+type Candidate = RunIxKeyIx
+
+-- | A candidate probe point is the combination of a filter\/run index,
+-- key index and hash number. This combination determines the probe location
+-- in the bloom filter, which is also cached.
+--
+-- We store these in 'PrimArray's as a pair of 64bit words:
+--
+-- * Low 64bit word:
+--   -    padding and the hash number (MSB)
```

Comment:
Yes and no. We don't need to mask out the low 16bits because we know the upper is empty. I'll see if I can clarify that.

Reply:
Sure, I just thought it would make it clear that we only use 16 bits for the hash number, since we assert that in `packCandidateProbe`. One might interpret the current comment as hallowing for any number of hash bits + some padding to get to 32 bits in total

Reply:
clarified